x.consolidation/consolidation.md

```diff
@@ -0,0 +1,517 @@
+almost always save more money than changeless transactions that require
+spending two or more UTXOs in the same transaction.  That's because
+transaction inputs (UXTO spends) for all commonly-used scripts are
+significantly larger than transaction outputs, so the savings of being
+able to omit an output doesn't compensate for the cost of having to
+include an extra input.
+
+| Type            | Input vbytes | Output vbytes | Ratio |
+|-----------------|--------------|---------------|-------|
+| P2PKH           | 148          | 34            | 23%   |
+| P2WPKH          | 68           | 31            | 46%   |
+| 2-of-3 P2SH     | 294          | 32            | 11%   |
+| 2-of-3 P2WSH    | 104          | 43            | 40%   |
+| P2TR (proposed) | 57           | 43            | 75%   |
+
+Changeless transactions also remove one UTXO from your wallet, reducing
+your future fee burden in the same way UTXO consolidation does.  The
+savings from this is harder to analyze because often you'll round up
+your fee in order to create the changeless transaction, increasing your
+spending feerate.  But even if we ignore that, inputs are so much larger
+than outputs that consolidating at a reasonably low rate will still
+save more money than creating two-input changeless transactions at a
+higher spending feerate.  Consider an example: Alice and Bob each
+receive 10 UTXOs.  Alice consolidates her UTXOs into a single UTXO at
+1/10th a normal spending feerate and then sends five payments.  Bob
+keeps his UTXOs separate and sends five payments of his own, each using
+two UTXOs to create a changeless transaction at the normal spending
+feerate:
+
+![Illustration of example consolidations and changeless transactions](img/cons-vs-changeless.png)
+
+Alice's Consolidation transaction with ten inputs and one output is
+679.25 vbytes; her subsequent five payment transactions each with one
+input and two outputs are 136.50 vbytes.  Bob's five changeless
+transactions, each with two inputs and one output, are 169.25 vbytes in size.  
+If Alice and Bob both send their payment transactions at the same rate,
+we can use some simple algebra to calculate what percentage of that
+feerate Bob needs to use for his consolidation transaction in order to
+save an equal amount of money as Alice saves using changeless
+transactions:
+
+    719.25 * x + 5 * 140.50 = 5 * 177.25
+    x = 0.26 (rounded)
+
+In short, if Alice can consolidate at a feerate of 1/4 of her regular
+spending rate or lower, she'll save more money than Bob even though she
+uses almost double as much block space as him.  The feerate data in the
+appendix to this chapter indicates such savings are possible under our
+rough assumptions over 96% of the time.
+
+If Bob needed to use more than two inputs per transaction in order to create
+a changeless transaction, Alice would be able to save more money even if she
+used a correspondingly-higher feerate for her consolidation transaction.
+
```

Comment:
I don't understand this comment.  In this example, Bob is using his UTXOs for changeless payments.

Reply:
Im trying to articulate that the chance of Bob even being able to execute a changeless payment due to amount mismatches (and having to overpay in order to achieve a changeless payment?) is probably low in practicality.

Reply:
In the provided example case where Bob only has 10 outputs to choose from (and they're all equal value), that absolutely is true.  In practice, a high-frequency spender who's also a high-frequency receiver will have a lot of UTXOs to choose from and those UTXOs may vary in value significantly, allowing them to take advantage of combinatorial blowup for finding good matches (or at least exponential increases if the spender is trying to limit their number of inputs to, e.g., 2).

For example, if Bob has `x` UTXOs each with a unique value, here's how many possible combinations there are of those UTXOs.  The more combinations, the more likely it is that Bob will find a combination that's close enough to his desired output value that he'll be able to produce a changeless transaction.

![2020-02-13-05_38_13_718656202](https://user-images.githubusercontent.com/61096/74430944-888d3980-4e23-11ea-8c3f-8aa04601db70.png)

In short, I'm not sure we want to imply that creating changeless transactions is unlikely, at least when applied to high-frequency spenders.  (For regular spenders who only have a few UTXOs at most, yeah it's unlikely.)

Reply:
Yep we are on the same page re: your explanation.

It may be worth adding clarifying language around the example in your text vs the "practical" example outlined in your comment above.